controller/src/main/java/io/pravega/controller/store/client/StoreClientFactory.java

```diff
@@ -115,14 +121,51 @@ public ZooKeeper newZooKeeper(String connectString, int sessionTimeout, Watcher
                 this.client = new ZooKeeper(connectString, sessionTimeout, watcher, canBeReadOnly);
                 return this.client;
             } else {
-                Preconditions.checkArgument(this.connectString.equals(connectString), "connectString differs");
-                Preconditions.checkArgument(this.sessionTimeout == sessionTimeout, "sessionTimeout differs");
-                Preconditions.checkArgument(this.canBeReadOnly == canBeReadOnly, "canBeReadOnly differs");
-                Preconditions.checkNotNull(watcher, "watcher");
-                this.client.register(watcher);
+                try {
+                    tryInjectFailure();
```

Comment:
One option here is to replace this `tryInjectFailure` call with:

```
Preconditions.checkArgument(this.injectParameterUpdateFailure.get() == true, "...");
```

I'm assuming this variable is an atomic boolean as suggested below, and we wouldn't need `tryInjectFailure` any longer.

I'm wondering if we can't instead mock curator so that we don't have to have any hack in the middle of the code. You are really trying to test two separate behaviors, which are of course connected, but still separate:

1- that `newZooKeeper` throws illegal argument in the case of a parameter change and closes the zk client
2- that the controller reacts correctly to `newZooKeeper` closing the client

If we separate the behavior like that, will it be simpler to test? For (1), we simply invoke `newZooKeeper` with a different parameter compared to the ones of the existing client. For (2), we close the zk client and check that the controller restarts correctly.

Reply:
I used an `AtomicBoolean` to inject an `IllegalArgumentException` in `StoreClientFactory` as suggested. And, as you correctly described, we are already testing both behaviors separately:

1- `StoreClientFactoryTest:testZkSessionExpiryWithChangedParameters`: This test verifies that in the case of an unexpected parameter change, we handle the exception correctly and close the zk client. Moreover, we also verify that further calls to Curator client return an exception that the Controller is able to handle correctly (`SessionExpiredException`).

2-`ZKControllerServiceMainTest:testZKSessionExpiry`: This test has been modified to also test that the Controller recovers correctly from a ZK session expiration in which ZK client parameters have been updated. This also involves the fact that the ZK client is closed.

For testing both scenarios, we added the `injectParameterUpdateFailure` flag (in a similar spirit to the `injectSessionExpiration()` method in ZK client). I feel that is convenient because it emulates the exact issue we are facing. Specifically, for test case number 2, we could close the zk client from the test code itself. However, we would not be in control of exactly "when" the client is closed: we may be closing it before or after the invocation to `newZooKeeper` upon a session expiration. This is because the invocation to `newZooKeeper` is internally managed by Curator client. This may lead the test to execute a different situation that is not exactly the one that we are facing in the original issue. I thought that exposing the `injectParameterUpdateFailure` flag is an acceptable way to have control over `StoreClientFactory` to reproduce the original issue with precision.